File: target_practice_prompts/swift/w3_d4_heapsort.swift
/**
 *  Target Practice 09 - Heapsort
 *
 *  NOTE: In-place means to manipulate the input array rather than create a
 *        new array.
 *
 *  Input:  arr: [Int]
 *  Output: [Int]
 *
 *  Example: heapsort([4, 15, 16, 50, 8, 23, 42, 108])
 *           //[4, 8, 15, 16, 23, 42, 50, 108]
 */

import Darwin

// Worse Time Complexity:
// Worse Auxiliary Space Complexity:
// Average Time Complexity:
// Average Auxiliary Space Complexity:

func heapsort(_ arr: inout [Int]) -> [Int] {
 // YOUR CODE HERE
 return [Int]()
}





////////////////////////////////////////////////////////////
///////////////  DO NOT TOUCH TEST BELOW!!!  ///////////////
////////////////////////////////////////////////////////////

func assert(_ count: inout [Int], _ name: String, _ test: () -> Bool) {
  if count.count != 2 {
    count = [0, 0]
  } else {
    count[1] = count[1] + 1
  }

  var pass: String = "false"

  if test() {
    pass = " true"
    count[0] = count[0] + 1
  }
  print(count[1], ")   ", pass, ":", name)
}

func arraysEqual(_ arr1: [Int], _ arr2: [Int]) -> Bool {
  if arr1.count != arr2.count {
    return false
  }
  for i in 0..<arr1.count {
    if arr1[i] != arr2[i] {
      return false
    }
  }
  return true
}


print("Heapsort Sort Tests")
var testCount : [Int] = [0,0]

assert(&testCount, "should sort [3,9,1,4,7]",  {
  var arr = [3,9,1,4,7]
  let test = heapsort(&arr)
  return arraysEqual(test, [1,3,4,7,9])
})

assert(&testCount, "should return empty array for empty input",  {
  var arr = [Int]()
  let test = heapsort(&arr)
  return arraysEqual(test, [])
})

assert(&testCount, "should sort single-element input",  {
  var arr = [10]
  let test = heapsort(&arr)
  return arraysEqual(test, [10])
})

assert(&testCount, "should sort moderate-sized input",  {
  var arr = [Int]()
  for i in 1..<1000 {
    arr.append(Int(arc4random_uniform(1000)))
  }
  var arr2 = arr
  let test = heapsort(&arr2)
  return arraysEqual(test, arr.sorted())
})

assert(&testCount, "should sort large input",  {
  var arr = [Int]()
  for i in 1..<1000000 {
    arr.append(Int(arc4random_uniform(1000000)))
  }
  var arr2 = arr
  let test = heapsort(&arr2)
  return arraysEqual(test, arr.sorted())
})

print("PASSED: ", testCount[0], " / ", testCount[1], "\n\n");

File: target_practice_prompts/cpp/w3_d4_heapsort.cpp
/**
 *  Target Practice 09 - Heapsort
 *
 *  Problem 1: Implement Heapsort in-place.
 *
 *  NOTE: In-place means to manipulate the input array rather than create a
 *        new array.
 *
 *  Input:  {Array}
 *  Output: {Array}
 *
 *  Example: heapsort([4, 15, 16, 50, 8, 23, 42, 108])
 *           //[4, 8, 15, 16, 23, 42, 50, 108]
 */

 // Worse Time Complexity:
 // Worse Auxiliary Space Complexity:
 // Average Time Complexity:
 // Average Auxiliary Space Complexity:

 #include <vector>
 #include <iostream>
 #include <algorithm> // for copy()
 #include <iterator>
 using namespace std;



 vector<int> heapsort(vector<int> arr) {
   // YOUR WORK HERE
   return arr;
 }



////////////////////////////////////////////////////////////
///////////////  DO NOT TOUCH TEST BELOW!!!  ///////////////
////////////////////////////////////////////////////////////

void runTest (bool test(), string testName, int testCount[]),  printTestsPassed(int testCount[]);
bool sortTest1(),sortTest2(),sortTest3(),sortTest4(),sortTest5();

int main() {
  int testCount[] = {0, 0};
  cout << "Heap Sort Tests" << endl;
  runTest(sortTest1, "should sort example input", testCount);
  runTest(sortTest2, "should return empty array for empty input", testCount);
  runTest(sortTest3, "should sort single-element input", testCount);
  runTest(sortTest4, "should sort moderate-sized input", testCount);
  runTest(sortTest5, "should sort large-sized input", testCount);
  printTestsPassed(testCount);
  return 0;
}

bool sortTest1() {
  return heapsort(vector<int>{3, 9, 1, 4, 7}) == (vector<int>{1, 3, 4, 7, 9});
}

bool sortTest2() {
  return heapsort(vector<int>{}) == (vector<int>{});
}

bool sortTest3() {
  return heapsort(vector<int>{10}) == (vector<int>{10});
}

// checks if array is sorted in linear time
bool isSorted(vector<int> input) {
    for (int i = 1 ; i < input.size() ; i++) {
        if (input[i-1] > input[i]) {
            return false;
        }
    }
    return true;
}

bool sortTest4() {
  vector<int> input(1000);
  vector<int> solution(1000);
  int randNum;
  srand(time(NULL));
  for (int i = 0 ; i < input.size() ; i++) {
    randNum = std::rand(); // rand() return a number between ​0​ and RAND_MAX
    input[i] =  randNum;
    solution[i] =  randNum;
  }
  input = heapsort(input);
  std::sort(solution.begin(), solution.end());
  return isSorted(input) && input == solution;
}

bool sortTest5() {
  vector<int> input(1000000);
  vector<int> solution(1000000);
  int randNum;
  srand(time(NULL));
  for (int i = 0 ; i < input.size() ; i++) {
    randNum = std::rand(); // rand() return a number between ​0​ and RAND_MAX
    input[i] =  randNum;
    solution[i] =  randNum;
  }
  input = heapsort(input);
  std::sort(solution.begin(), solution.end());
  return isSorted(input) && input == solution;
}


// this is to wrap the test and check for exceptions
void runTest (bool test(), string testName, int testCount[]){
  testCount[1]++;
  bool testPassed = test();
  if(testPassed) testCount[0]++;
  string result = "  " + (to_string(testCount[1]) + ")   ") + (testPassed ? "true" : "false") + " : " + testName;
  cout << result << endl;
}

// this is to print the number of test cases passed for a test suite function
void printTestsPassed(int testCount[]) {
  cout << "PASSED: " + to_string(testCount[0]) + " / " + to_string(testCount[1]) + "\n\n" <<endl;
}
